docs/internal/expressions.md

```diff
@@ -0,0 +1,611 @@
+
+### Structural equality
+
+Structural equality, compared with the method `==` on appropriate types, is a comparison of the values in the object that the variable points to, including numeric literal values.
+It is the responsibility of each class to implement their own structural equality method.
+
+User objects can be compared structurally, simply by implementing the `==` method.
+Similar operators (`!=`, `<`, `>`, etc) are also implemented in the standard library and can be used by any user class in the same way.
+
+Example:
+```ts
+// Some variable
+var a : U32 & imm = 42;
+// A new variable with different value
+let b = a + 1;
+// A new variable with the same value (but different address)
+var c : U32 & imm = 42;
+
+let ne = (a == b); // FALSE, 42 != 43
+let eq = (a == c); // TRUE, both as 42
+```
+
+_Note: There is nothing special about structural comparison methods.
+ They're methods like any other that just happens to "implement" comparisons.
+ You can do whatever you want with your own types for `==`._
+
+## Syntax Sugar
+
+In Verona, operations are method calls.
+To create syntactically correct code, one must be explicit about their names and types (arguments and return values).
+
+Example:
+```ts
+class MyType {
+  create() : MyType;
+  apply();
+}
+// Creates a new object and calls it
+let a : MyType & mut = MyType::create();
+a.apply();
+```
+
+To make the code above more readable, Verona introduces _syntax sugar_, which allows you to write more concise code meaning the same thing.
+
+The two main sugars are:
+* `create`, for constructing objects, allowing you to call just the class name instead of the whole call syntax.
+* `apply`, for calling objects as functions, including lambdas, without the need to call it directly.
+
+The code above would look like:
+```ts
+let a : MyType & mut = MyType;
+a();
+```
+_Note: We also want the `update` sugar, but it may need a `Ref[T]` type to do `obj.update(field) = val` where `update()` returns `Ref[T]`._
+
+## Object Literals
+
+Object literals are anonymous types created by the declaration of an object with fields and methods.
+The types of the fields and methods are declared or inferred, and the final type is created within the context that it's used.
+
+The syntax for object literals is still under discussion.
+
+Lambdas are an example of object literals, where the fields are the captures (and their types inferred from the objects captured) and the single method `apply()` is the body of the lambda.
+
+## Partial Application
+
+Partial application refers to the process of fixing some arguments of a call and returning a lambda with the remaining arguments to be passed to the lambda.
+As most of Verona's high-level functionality, it will be implemented in the Verona standard library as a method called `~`, using type lists for the arguments.
+
+TODO: Add an example here.
```

Comment:
Partial application is purely a standard-library thing, so we can probably drop it here.